### PR TITLE
Support GHC 8.1.

### PR DESCRIPTION
**Pull request by svenpanne**
It would be great to have a release soon, currently all Travis-CI jobs with GHC HEAD which use ``hashable`` are broken.

### CODE REVIEW

**Comment by phadej** [comment]
@svenpanne use `--allow-newer` (or `hashable` specific versions of that flag). See e.g. https://github.com/ekmett/bytes/blob/4b933f8b650b29567b684d67ac8c2979776a0895/.travis.yml#L31

IMHO it's a bit too early to push GHC 8.2 stuff to Hackage.

**Comment by svenpanne** [comment]
I don't understand what's the point of *not* relaxing the upper bound and releasing a new version: ``hashable`` compiles and its tests work. Telling all clients of ``hashable`` to use the ``--allow-newer`` sledgehammer instead of doing the real stuff upstream is a bit lazy and somehow contradicts the intention of version bounds.

**Comment by phadej** [comment]
@svenpanne you can say (in `cabal`) `--allow-newer=hashable`. 

Also `hashable` compiles and tests works /now/, the 8.1 isn't frozen yet. It's unluckily to broke, but it's not impossible.

Also `deepseq` introduces `NFData1`, which `Hashed` could implement if @tibbe decides to include `NFData Hashed`, but it's not yet clear what the next version of `deepseq` would be (e.g. current `primitive` on Hackage fails to build).

It's not fair to ask maintainers to keep their libraries up to date with the unreleased bleeding edge.

**Comment by svenpanne** [comment]
I think asking maintainers to keep their libraries up to date with the unreleased bleeding edge is the only way to get a smooth transition in the package ecosystem when a new GHC comes out: At that point, you definitely *don't* want to use ``--allow-newer``, otherwise you could just remove your upper version bound. So you are basically forced to wait until all packages up the dependency chain have updated their version bounds and do a release, which can take quite a long time if you have a lot of dependencies. If package maintainers release proactively, this is much less painful. ``hashable`` itself has very few dependencies, so it is not affected very much by this problem, but think about other packages much farther down the dependency chain. Another option would be using less strict version bounds, but that has other implications.

In a nutshell: Having very strict version bounds and not updating/releasing proactively does harm to the whole package ecosystem, but that's just my opinion...

**Comment by phadej** [comment]
@svenpanne I agree that it's good to prepare for GHC release, but can't we wait till the first (which won't be the last) release candidate?

**Comment by svenpanne** [comment]
That would be OK, I guess: https://ghc.haskell.org/trac/ghc/wiki/Status/GHC-8.2.1 says that a RC should be available in a few weeks, and we all have a time window of about 2 months to fix/update packages then.

**Comment by tibbe** [comment]
As you both seem to agree, there's a trade-off between
* getting hashable working with an unreleased GHC, to reduce latency between the GHC release, the hashable release, and the release of dependent packages, and
* avoiding duplicating work in case of GHC changes before release.

There's also a difference between having hashable HEAD working with the unreleased GHC, so people can test it using e.g. Stack, and having a released version of hashable that does. The latter can be dangerous because late changes to GHC could invalidate the Hackage release (e.g. it might not build, etc.) and we can't withdraw releases (although we could perhaps update the Hackage metadata so the package doesn't get picked up).

I wish the Haskell community would be slower to pick up new compilers (like many other communities) and we could use 1-2 months after the GHC release to get everything working, but that's not the case.

I'm OK making the compromise that once the GHC RC is out, we can try to make a release.

**Comment by RyanGlScott** [comment]
This was resolved in 4b847ef6f358f66119cf102991d07a3588c889db.
